src/odb/src/3dblox/chipletHierarchy.cpp

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2019-2025, The OpenROAD Authors
+
+#include "chipletHierarchy.h"
+
+#include <algorithm>
+#include <queue>
+
+#include "utl/Logger.h"
+
+namespace odb {
+
+ChipletHierarchyAnalyzer::ChipletHierarchyAnalyzer(utl::Logger* logger)
+    : logger_(logger)
+{
+}
+
+std::vector<ChipletLevel> ChipletHierarchyAnalyzer::analyzeHierarchy(
+    odb::dbDatabase* db)
+{
+  levels_.clear();
+  chiplet_levels_.clear();
+
+  // Build dependency graph
+  auto dependencies = buildDependencyGraph(db);
+
+  // Perform topological sort to determine levels
+  performTopologicalSort(dependencies);
+
+  // Convert to vector format
+  std::vector<ChipletLevel> result;
+  for (const auto& [level, chiplets] : levels_) {
+    ChipletLevel chiplet_level(level);
+    chiplet_level.chiplets = chiplets;
+    result.push_back(chiplet_level);
+  }
+
+  // Sort by level number
+  std::sort(result.begin(),
+            result.end(),
+            [](const ChipletLevel& a, const ChipletLevel& b) {
+              return a.level < b.level;
+            });
+
+  return result;
+}
+
+std::vector<dbChip*> ChipletHierarchyAnalyzer::getChipletsAtLevel(int level)
+{
+  auto it = levels_.find(level);
+  if (it != levels_.end()) {
+    return it->second;
+  }
+  return {};
+}
+
+int ChipletHierarchyAnalyzer::getChipletLevel(dbChip* chiplet)
+{
+  auto it = chiplet_levels_.find(chiplet);
+  if (it != chiplet_levels_.end()) {
+    return it->second;
+  }
+  return -1;
+}
+
+std::map<dbChip*, std::set<dbChip*>>
+ChipletHierarchyAnalyzer::buildDependencyGraph(odb::dbDatabase* db)
+{
+  std::map<dbChip*, std::set<dbChip*>> dependencies;
+
+  // Initialize all chiplets with empty dependency sets
+  for (auto chiplet : db->getChips()) {
+    dependencies[chiplet] = std::set<dbChip*>();
+  }
+
+  // Find dependencies by analyzing chiplet instances
+  for (auto chiplet : db->getChips()) {
+    for (auto inst : chiplet->getChipInsts()) {
+      auto master_chip = inst->getMasterChip();
+      if (master_chip != nullptr) {
+        // The chiplet depends on its master chip
+        dependencies[chiplet].insert(master_chip);
+      }
+    }
+  }
```

Comment:
Merge the 2 for loops, there is no need to split. Also master chip can never be null.
```suggestion
  for (auto chiplet : db->getChips()) {
    for (auto inst : chiplet->getChipInsts()) {
      auto master_chip = inst->getMasterChip();
      dependencies[chiplet].insert(master_chip);
    }
  }
```